### ISSUE TITLE
Bitfunx memcardpro+ v1.0.3 with Game ID support issue with LastMemcardIndex.dat.

### ISSUE THREAD

**Issue by whitezombie2000** [MrPalloncini/PicoMemcard#1]
Using the Xstation with an NTSC SCPH-5501, after a power cycle, or if you unplug the card during gameplay, it will not load the last memory card used. It doesn't seem to honor the LastMemcardIndex.dat. It will keep starting memory card 9 if you go to the memory card manager or if you unplug the memory card during gameplay. This doesn't affect launching a new game session as it will load your memory card if it exists in the index.txt file.  

If I delete memory card 9 then it will default to memory card 1 upon a new power cycle when going to the memory card manager or reinserting during gameplay. I had about 20 memory cards created when testing.

The fork by realJoshByrnes doesn't have this issue (doesn't have game ID support but implemented LastMemcardIndex.dat), and will always honor the LastMemcardIndex.dat after a power cycle, or if you unplug it during gameplay, and will load the last memory card used.

https://github.com/dangiu/PicoMemcard/issues/40#issuecomment-1482388229

![Capture1](https://github.com/MrPalloncini/PicoMemcard/assets/23583562/ae6cc37f-6e65-4c12-b3cb-414fd37e20f1)

**Comment by MrPalloncini** [comment]
This need a lot of debugging, i think i will need to reimement everything inside the index.txt file and also rewrite the enitre cose that order the memcards in decimal order instead of alphabetical

**Comment by whitezombie2000** [comment]
No worries, appreciate your work! But yeah, LastMemcardIndex.dat doesn't update when switching memory cards. It will update if you create a brand new card but won't switch if you change to an existing card. If you manually change the number in LastMemcardIndex.dat by opening the file in notepad, saving it, it will load a different card, but it will not load the correct number card due to the firmware sorting alphabetical instead of decimal.

**Comment by MrPalloncini** [comment]
Fixed in the next release, i've dropped the .dat file and everything will be inside the index.txt, make sure the first line will be LASTMEMCARD:X.MCR. Also i dropped the ordering since it wasn't needed. now the application knows the position based on the loaded mcr file, which of course is faster and better. If there are any more issues, feel free to open again or open a new one after i post the new release

**Comment by whitezombie2000** [comment]
Great work on the latest update v1.0.4! I tested a PSIO and an X-Station. The creation of new cards, switching, and last LASTMEMCARD:X.MCR worked great. However, something broke with Unirom. I tested FreePSXBoot and a cartridge flashed with Unirom 8.0.K.

Unirom is creating new cards every time at the Unirom menu. It is not detecting existing Game ID saves and is not creating new Game ID saves.

Here's a screenshot before I started testing Unirom and was only testing the ODEs.

![Capture0](https://github.com/MrPalloncini/PicoMemcard/assets/23583562/71d86e8b-0431-451c-bfc3-84bfe79580a2)

Here's after trying FreePSXBoot slot 2 and a flashed cartridge a couple of times each. They both created two new memory cards, 37, 38, 39, and 40.
![Capture1](https://github.com/MrPalloncini/PicoMemcard/assets/23583562/001ee4fc-6708-471c-ae45-7bef303d5e96)

**Comment by MrPalloncini** [comment]
My bad, i should have tested it with Unirom, I'll see what i can do about it
EDIT: I can't quite track the issue with unirom, seems like that the way it's sending the gameID is quite strange, the application can't keep up with the data it sends. This will require way more testing and to be honest, the entire application needs a new rewrite for the operation and basecally emulate memcardpro for a full good compatibility, right now due to my low exprerties on programming and embedded devices i can't do that, i'm sorry. Until this issue is fixed i would say that Unirom is not supported.

**Comment by whitezombie2000** [comment]
Thanks for your effort, I'll stick to the previous release v1.0.3 for now as Unirom, PSIO, and X-Station was working with Game ID. I pinged dangiu and realJoshByrnes to see if they can help.

**Comment by whitezombie2000** [comment]
I also pinged Rama to see if he can help as he stated Unirom and X-Station are the same mechanism. 

https://github.com/dangiu/PicoMemcard/issues/61#issuecomment-1612189156

**Comment by MrPalloncini** [comment]
That's why i can't figure out why Unirom is buggy, because the code works ok with xstation and psio (according to the users), i don't know why the gameid variable is not passed completely to the index function. Also the application fails to get the entire gameid sometimes from unirom. My fear is that if i make the code function with unirom it will be bugged with the ode like before. I'm sorry but i'm still learning C programming as i never coded something like this big before, i don't want to let down the community so i apologize, but right now i want the code to work with those odes since the great part of the community is what they are most likely to use. Any help will be appreciated and if a fix is found i will do my best

**Comment by whitezombie2000** [comment]
No worries, you did a great job so far! It's weird that Game ID and the index.txt works with Unirom on v1.0.3. I wonder what drastically changed with the v1.0.4 build.

**Comment by MrPalloncini** [comment]
Does the pre release firmware have this bug?

**Comment by whitezombie2000** [comment]
Unirom on v1.0.3 and v1.0.3 pre-release functions just like post #1 with the X-Station. It will create new cards if the game ID does not exist, it will update the index.txt with a game ID and card #, it will load an existing game ID memory card just fine when you start a game, it will update the LastMemcardIndex.dat file upon creation of a new card only, and upon switching a game that exists in the index.txt, the LastMemcardIndex.dat file is not updated.

Edit: So with v1.0.3 pre-release I just tested switching memory cards with START + SELECT + DPAD UP and START + SELECT + DPAD DOWN at the CD PLAYER\MEMORY CARD screen. When you use the button combo the LastMemcardIndex.dat is updated but will load the incorrect card if you go back to the CD PLAYER\MEMORY CARD screen after a power cycle due to the alphabetical\decimal issue. The PSIO, X-Station, and Unirom do not update the LastMemcardIndex.dat when you load an existing game in the index.txt file, but will update the LastMemcardIndex.dat when you create a new memory card, and new game ID in the index.txt file. Also, the LastMemcardIndex.dat file will update if you create a new generic card with START + SELECT + TRIANGLE which is a normal function.

In conclusion, the PSIO, X-Station, and Unirom were all functioning the same on v1.0.3 pre-release with my testing.

**Comment by whitezombie2000** [comment]
Also, have you considered using AI like ChatGPT to assist with the code? I have no idea on how to code or compile, I took the contents of memcard_manager.c, pasted it into ChatGPT, it then gave a description of what the code is doing. Then I typed in "Fix the code to sort the .MCR files by decimal". It then wrote out code modification and an explanation at the end. Maybe consider registering an account, and try it out, try different phrases like"Fix the code to sort the .MCR files numerically", maybe it can help you improve the code so we can get it working since you are new to the language. Let me know if it works out as I can do some testing. I would be happy to help!

https://chatgpt.com/share/947172c5-e0b3-4340-8b00-779007ec596f

https://chatgpt.com/share/3590c760-684d-4a3c-be3a-5d412b34bb3b

**Comment by MrPalloncini** [comment]
sorting is not required anymore since i can track the active mcr file and then know which one i should pick. I don't even know why the sorting was there to begin with. I might have patched some issues with unirom, but i don't have too much time to test it
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16236983/PicoMemcard.uf2.zip)
There might be a bug with powercycles. if the console resets too quickly the app basecally goes nuts. If this happen, do a full powercycle. Reports to me every finds. Sorry if i was late to this and i can't do much testing

**Comment by whitezombie2000** [comment]
The good news is that Unirom stopped creating new cards at the menu on every boot. The bad news is that everything else is very buggy.

I tested three games with Unirom, Resident Evil 3, Resident Evil 1 DC, and Final Fantasy Tactics. First test it created SLUS_009.23:36.MCR in the index.txt but failed to create a 36.MCR card. Second game I booted created SLUS_005.51:36.MCR and created a 36.MCR.

![Capture3](https://github.com/user-attachments/assets/442f3bfd-56a3-4291-975b-a11a3bf2f58f)


I deleted those two, tried again, it created the two without issue. 

![Capture4](https://github.com/user-attachments/assets/a10b89d8-034d-4cb6-a17a-288dbaf84a86)

Resident Evil 1 DC SLUS_005.51 for some reason using Unirom and X-Station creates the entry without a \ if you noticed in the pics above. This seem to happen on v1.0.3 pre-release too. If I manually put the \ in there it will ignore it and load the one without the \ . PSIO creates the entry normally as \SLUS_005.51. Sometimes it would also delete the \ from the first entry \SLUS_007.26:26.MCR.

![Capture5](https://github.com/user-attachments/assets/36c8b0a2-6a10-462b-94da-463ec7753fce)

Also, one Unirom test it deleted the first entry \SLUS_007.26:26.MCR and added a second LASTMEMCARD entry.

![Capture6](https://github.com/user-attachments/assets/da98b6c7-b0ae-45bb-9d2e-39b33088d945)

On the X-Station and PSIO it was creating the last three entries like this.

![Capture7](https://github.com/user-attachments/assets/1a55bb7a-baf3-4855-ab9b-6ef2a06801da)

X-Station and Unirom v1.0.3 pre-release (and later releases) with Resident Evil 1 DC SLUS_005.51. I don't know if other games do this, too many to test!
![Capture8](https://github.com/user-attachments/assets/42c3e745-a917-4c22-b6fe-ebc67a8617ba)


Yeah, it's up to you what you want to do at this point. Keep debugging your current builds and try to fix all these issues. Or drop back to v1.0.3 pre-release with LastMemcardIndex.dat, and implement the minor game fixes from v1.0.4 into v1.0.3 pre-release, because PSIO, X-Station, and Unirom worked great with v1.0.3 pre-release. I think the sorting was there for LastMemcardIndex.dat, I could be wrong on that though, I guess only dangui can really answer that. Or maybe the sorting was required for START + SELECT + DPAD UP and START + SELECT + DPAD DOWN manual switching order. I am here to help in any way I can to make this work, would be nice if we had some others testing, lol, takes a lot of time to test all three platforms, I'm going to wear out the USB C port on the card, ha ha! :)

**Comment by ramapcsx2** [comment]
Couple ideas:
Make sure to disregard one of the (possibly) 2 ID messages. There is one sent per controller port, but that may not always be the case.. Just make sure that when you got one, to ignore the one from the other port.

Next, the way these lists come out looks like sometimes the string terminator is ignored. String handling in C is a bit tricky, you can ask ChatGPT whether your buffer arrays will be able to hold complete file names and whether any code that deals with the buffer arrays has a "off by one" error. It happens all the time, super common.

**Comment by MrPalloncini** [comment]
The array sent doesn't have a termination character and the len of the string is bigger by 1. Of course i decrease the lenght and the null terminator character is present in the final array

**Comment by ramapcsx2** [comment]
If you can reproduce the error, and have a serial log connection, try printing out the exact file names to terminal, see if it pops up there.

**Comment by MrPalloncini** [comment]
I will test by rolling back the changes and then realeasing a functional release. Untill then who uses unirom please use the 1.0.3, sorry for the trouble, i'll do my best to patch this as soon as possible, due to my real life problems i don't know when i will publish the new update.

**Comment by MrPalloncini** [comment]
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16321225/PicoMemcard.uf2.zip)
Please test this file and tell me if anything changes, there are still problems with resets i think

**Comment by whitezombie2000** [comment]
The file you uploaded is an elf. I'm guessing it needs to be converted to a uf2.

**Comment by MrPalloncini** [comment]
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16322082/PicoMemcard.uf2.zip)
sorry i got distracted, i'm a bit tired those days and i'm trying my best to rush and fix those issues. I'm glad you can help me to test things better since any minor change can effect the outcome of the application

**Comment by whitezombie2000** [comment]
Pretty good job with this build as it seems much more stable. I briefly tested the three same games with Unirom, X-Station, and PSIO. It created all three without issue. Resident Evil Director's Cut on Unirom and X-Station still creates the entry without the backslash, as SLUS_005.5, and PSIO creates this game properly as \SLUS_005.5. It doesn't really seem to affect anything. Again, not sure if other games have this issue as we don't have extensive testing.

![Capture1](https://github.com/user-attachments/assets/fcdf13c7-ed22-49f8-9149-217c9d3c11e7)

There still is a minor issue where it will sometimes improperly place the new entry in the index.txt and will not recognize the card because it will mess up the file name. This can randomly happen on Unirom, PSIO, and X-Station when creating a new game ID. This can be remedied by manually moving the file name down one row. The below screenshot is from the X-Station where I booted, reset, and created 10 new cards. On the very first try it improperly created the entry.

![Capture2](https://github.com/user-attachments/assets/6f7ab286-ad14-4b75-aafb-19e56d2a764a)

**Comment by whitezombie2000** [comment]
I just figured something out, there is a new blank line in the index.txt file at the end of every entry, and I didn't notice it before. I believe that is why some of my game ID entries were being entered in wrong. In my previous post where I tested 10 games on the X-Station, I entered LASTMEMCARD:0.MCR and never had a new blank line entry. That is why it entered the first entry incorrectly as LASTMEMCARD:0.MCR\SLUS_005.53:1.MCR. 

This morning I created 20 new cards with the PSIO. As a test I removed the blank line on the last entry and it put the entry in wrong when creating the next card.

![Capture1](https://github.com/user-attachments/assets/71f60a08-43ed-4ec4-a29b-3e769a787b68)

I tested it one more time with the blank line entry and it was created properly. So I guess if you edit the index.txt file you need to press enter at the end of a manual edit to make a new blank line, or make sure it exists, don't press backspace, or delete the blank line entry.

![Capture2](https://github.com/user-attachments/assets/45373e77-64aa-4b14-8c34-6415d2520d48)

**Comment by MrPalloncini** [comment]
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16325798/PicoMemcard.uf2.zip)
Ok so... here what should be fixed now:
UNIROM should now correctly send the game id (i hope)
a new line is automatically added (no need to add it yourself so try recreating the index file)
let me know if you find more issues

**Comment by whitezombie2000** [comment]
I will test it when I have the opportunity. It probably won't be until later this week. Just so you are aware the previous update you made was creating a new line in the index.txt. I was accidently removing it because I didn't realize it was needed.

**Comment by whitezombie2000** [comment]
Hey, I had a small window to test before I start going back to work for the week. This release broke Unirom and will not create memory cards. I briefly tested two games each on PSIO, X-Station, and Unirom. You need to fall back to your previous release as it was stable with all three. Do you have a device that you can test Unirom with? I understand if you don't but it would be nice if you could test it before uploading it. Again, I understand if you don't have a working CD drive. Have a good week!


Edit 1: Ok, I quickly tested three games with Unirom again. This time it created all three without issue. Not sure what happened the first time but worked on the second test. Maybe later in the week I can do some more testing but I'm out of time for now! I wish we had few more people to help us! See you later!

![Capture1](https://github.com/user-attachments/assets/34125036-37ae-4924-ba65-5487bdfa1afc)

Edit 2: It does appear to work with Unirom but some times it doesn't create the card and game ID. Just did another quick test with 8 CDs. The first four created without issue using a full power cycle off and back on. But when I tried the 5 CD it didn't work, powered off and back on, failed again, powered off and back on, worked on the third attempt. 6th, 7th, and 8th CD had no issue creating cards.

![Capture2](https://github.com/user-attachments/assets/d9b76918-3575-4907-b40f-dab5003d1b9d)

Edit 3: There are still bugs. I created 10 saves on the X-Station and at some point it deleted the first entry letter "S".

![Capture3](https://github.com/user-attachments/assets/493b1125-66d2-42bf-90d0-43793a025d51)

Created 10 saves on the PSIO.

![Capture4](https://github.com/user-attachments/assets/084037e0-fb44-434b-afb7-4b981d8d6680)

**Comment by MrPalloncini** [comment]
i was testing with uniboot, i thought that the unirom cd would be the same thing

**Comment by whitezombie2000** [comment]
Never heard of Uniboot or used it. Probably doesn't matter if it's a burned CD, FreePSXBoot, or a flashed cartridge, as long as it's the latest UniROM 8.0.K. The majority of my testing has been with FreePSXBoot slot 2, freepsxboot-unirom-fastload-20220919-bios-4.4-2000-03-24-A-6a0e22a0-slot2.

**Comment by MrPalloncini** [comment]
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16373972/PicoMemcard.uf2.zip)
ok, try this version, i've made a bunch of fixes, still not perfect since i have to refine few things, let me know what you find

**Comment by whitezombie2000** [comment]
Awesome work! Did some testing for a few hours with Unirom, PSIO, and X-Station. This seems pretty stable, could even be a stable release, but did encounter one minor bug.  Above I posted about the first memory card entry letter ''S'' being deleted with the X-Station. I have seen this bug with all the releases and I think it happened on all three platforms. After creating all the new entries with Unriom and no issues, I booted a few more times, tested some game saves, manually switched some cards with the button combo, put the card back in the PC, and the first memory card entry letter ''S'' was deleted. I put the "S" back in, saved the index.txt file, tried a dozen restarts, dozen power cycles, game loads, some write operations, new memory card entries, and could not reproduce it.  Also, retried with X-Station, and PSIO to re-create this bug, but could not trigger it. I don't know when this bug is being triggered, it's a very minor one, but it's tough to figure out when it happens.

Tested 11 games with Unirom and a Game Shark 4.0 CD (ARCDX.EXE). PSone NTSC SCPH-101  was used, powered off, and powered on for all games.

![Capture1](https://github.com/user-attachments/assets/efe22980-5ea8-4b09-9ca2-3df4dd6612be)

Tested 12 games with PSIO. NTSC SCPH-5501 was used and reset between each game.

![Capture2](https://github.com/user-attachments/assets/22e39c5e-c8d7-4a14-a8bc-e39a9ecbce25)

Tested 13 games with X-Station.  NTSC SCPH-5501 was used and reset between each game.

![Capture3](https://github.com/user-attachments/assets/ee0e316d-dc8e-456c-94c6-a84193199f8a)

**Comment by whitezombie2000** [comment]
OK, I have an update. I can reproduce the first memory card name entry issue. Here's how to reproduce it. In the index.txt file put a single digit entry like 0 through 9 in LASTMEMCARD: and save it . When there is single digit entry, put the card into the PS1, manually create a new memory card with the button combo SELECT+START+TRIANGLE. It will delete the letter every time in the first entry. If I put a double digit entry in there like 10 through 20 it never seems to happen with the button combo SELECT+START+TRIANGLE. I can also reproduce the issue with ODEs and Unirom when creating a new game ID entry but I can't reproduce it every time with a single digit in LASTMEMCARD: . It seems to be very infrequent on the ODEs and with Unirom when you create a new game ID. I hope this helps you to debug!

![Capture4](https://github.com/user-attachments/assets/2455c63a-25a1-46c3-9eb3-583cbaa950e7)

**Comment by MrPalloncini** [comment]
nice catch and nice debug man, you are the best tester i've ever seen. I'll see what i can do about it!

**Comment by whitezombie2000** [comment]
If you fix this (and don't introduce more bugs, lol) I think we may have a very stable build that you can release.

**Comment by ramapcsx2** [comment]
Working with file systems is not easy, just wanted to mention that. If you fix it all, that's quite good work :)

**Comment by whitezombie2000** [comment]
Maybe an easy fix would be to have the number entries as, 00, 01, 02, 03, ect. Let me know what you find out and works best. I might be able to test it out next weekend. Currently enjoying the Olympics so have limited time to do testing.

**Comment by MrPalloncini** [comment]
does switching memcard from 9 to 10 triggers this bug?

**Comment by whitezombie2000** [comment]
I'm not 100% positive on your question. I think it only triggers when a new memory card is created. I don't think it happens when switching but you can certainly test switching between them to see if it triggers.

**Comment by MrPalloncini** [comment]
yep, it does, the issue here is that when a number greater than 2 character is written is overwriting the strings that follows, i might have an idea to fix that but the code would be ugly.
EDIT: Actually it wasn't that bad, i tested and it should work, from now on the string must be LASTMEMCARD:009.MCR, so it will always overwrite itself and not the other strings (managing strings in c is a pain). so recreate your index file or just edit it that way.
[PicoMemcard.uf2.zip](https://github.com/user-attachments/files/16404879/PicoMemcard.uf2.zip)

EDIT2: i've just fixed the last bug i had on the schedule so, if this version does not have more bugs, i will release a stable one after few more testings

**Comment by whitezombie2000** [comment]
Amazing work! Tested out all three platforms and had no issues!

**Comment by MrPalloncini** [comment]
Thank you, i added you to the credits for your awesome work with the report and debug

**Comment by Playmakeree7** [comment]
Hello, I have a question, how modify the index.txt file to keep FreePSXBoot as the default image during startup using the MrPalloncini/PicoMemcard firmware? I read this at amazon:

"To start, I did have some issues with the dangiu/PicoMemcard firmware update that was originally suggested by Bitfunx. The Bitfunx team was very responsive and suggested I try the MrPalloncini/PicoMemcard firmware, which is much improved in a number of ways. There is one caveat, being that if you use FreePSXBoot you will also need to modify a line on the index.txt file, which is honestly extremely easy to do on a PC. But, let me get back to the PSxMemCard itself.

Bonus features if you use the MrPalloncini firmware and FreePSXBoot+Unirom --
This combo adds the feature of automatically creating and switching to virtual memory cards on boot depending on the game CD. You may need to modify the index.txt file as mentioned earlier to keep FreePSXBoot as the default image during startup, but after that it's been smooth sailing. No need for manual card swapping, and every game gets its own card. Life is good! :)"

**Comment by whitezombie2000** [comment]
@Playmakeree7 
Use sd2psXtd firmware here, https://github.com/sd2psXtd/firmware, grab the psxmemcard.uf2 file, and flash it. The sd2psXtd firmware is probably the future for these cards, I say that with no disrespect to MrPalloncini as he did a great job, I enjoyed helping him, but sd2psXtd with MMCE has a large community around it. You can even use these as PS2 memory cards, and play PS2 ISOs from them on your PS2. The MMCE group includes (but is not limited to) 8BitMods and sd2psXtd developers. It's really great so give it a try.


Set your settings.ini file like this and place your FreePSXBoot image in SDCard:\MemoryCards\PS1\Card1\Card1-1.mcd .

[General]
Mode=PS1
FlippedScreen=OFF
[PS1]
Autoboot=OFF
GameID=ON
[PS2]
Autoboot=OFF
GameID=ON
CardSize=8
Variant=RETAIL

**Comment by Playmakeree7** [comment]
Thanks @whitezombie2000  it works fine with unirom, except with tonyhax international for APv2 games. When using tonyhax, the memory card doesn't switch.
I'm happy with unirom, thanks for the help, best regards.
